### ISSUE TITLE
Discuss WAMP-CRA information leakage

### ISSUE THREAD

**Issue by mbonneau** [wamp-proto/wamp-proto#100]
In the [WAMP-CRA section](https://github.com/tavendo/WAMP/blob/master/spec/advanced.md#wamp-challenge-response-authentication) of the advanced spec, it looks like the user's role is exposed prior to user authentication.

This creates opportunity for malicious users to fish for an account with elevated access.

I think it would be good to drop the authrole from the challenge.

**Comment by konsultaner** [comment]
I was about to open the same issue as @mbonneau. It seems like it is possible to check whether a userid exists or not, especially when you have many roles. The problem is that it is unclear to decide what role to expose if the authid does not exist at all without telling everyone that it actually does not exist.

**Comment by gammazero** [comment]
@mbonneau Removing authrole from the challenge is not sufficient in some cases.  Consider that the presence or absence of salting info may also reveal whether or not an authid is valid.

What is currently leaked is:
- Whether or not a user (authid) exists or not.
- Which user has which roles
- Whether or not PBKDF2 key is used

If an authid does not exist, authrole could be set to "user" in the challenge to make it a little harder to tell if the authid exists or not.  However, this is still weak, since as @konsultaner said is a problem especially when you have many auth roles. So challenges with authrole values that are not "user" give away a valid authid.  Also, whether or not salting information appears in the challenge can also give away the authid.

For information not to be leaked, a challenge issued for an invalid authid must be indistinguishable from a challenge issued for a valid authid.  This does not appear to be possible given the intent of CR auth.  If the information leakage is not acceptable, then a different auth method should be chosen.

Personally, I like to handle auth outside of WAMP - use client certificates + TLS, then the router only needs to be concerned about authz. In general, I think wamp should try to stay out of the business of defining authentication methods and stick with a minimal set of functionality that lets it integrate with existing authentication systems.

**Comment by oberstet** [comment]
Everyone, and everything should use TLS always. If you do not run TLS, then what's the point of protecting the auth handshake from leaking, while leaving the actual app traffic exposed?

**Comment by mbonneau** [comment]
@oberstet TLS would not fix the information leakage issue unless you are using certificates for authentication prior to the WAMP handshake. You could still connect with TLS and fish for information.

**Comment by oberstet** [comment]
ahh, right. so this is not leaking to a man-in-the-middle, but a bad guy just trying to connect.

ok, yes. we could remove `authrole` from the challenge to stop that leaking.

but we would also need to specify that a router would _always_ challenge the client, even when it already knows it is not accepted (eg `authid` is not know).

**Comment by gammazero** [comment]
@oberstet What about the presence or absence of salting information in the challenge?  That could also give away whether the authid existed.

**Comment by meejah** [comment]
@gammazero you have to be even more careful than just "is there salt", it has to be the same for the same (wrong) authid -- if I try twice and get different salt information back, I know that's not a "real" user.

**Comment by meejah** [comment]
> but we would also need to specify that a router would always challenge the client, even when it already knows it is not accepted (eg authid is not know).

Yes, this is the only way to prevent this information-leakage -- the behavior of the router must be identical regardless of whether an authid (and/or authrole?) exists.

This implies you also have to "remember" the fake information you sent out; for WAMP-CRA and SCRAM, you'd have to remember the salts (one way to "remember" without wasting unbounded resources would be to keep an HMAC key for this purpose, and use the HMAC for the wrong authid+some string as the salt). For wamp-cryptosign it's easy (but you do still have to give them a challenge AND CHECK IT when it comes back, or the timing will leak the existence of the authid).

**Comment by oberstet** [comment]
> For wamp-cryptosign it's easy (but you do still have to give them a challenge AND CHECK IT when it comes back, or the timing will leak the existence of the authid).

Huh, yeah, timing attacks. Good point!

We could make WAMP-cryptosign fixed run-time:

* if fake, just sleep for N ms
* if not fake, check (which runs in M ms), and then sleep for N-M ms
